File: src/p0002/cpp/solution.cpp
class Solution {
public:
    static ListNode *addTwoNumbers(ListNode *a, ListNode *b) {
        ListNode sentinel(0), *tail = &sentinel;
        for (int carry = 0; carry || a || b; carry /= 10) {
            if (a) {
                carry += a->val;
                a = a->next;
            }
            if (b) {
                carry += b->val;
                b = b->next;
            }
            tail = tail->next = new ListNode(carry % 10);
        }
        return sentinel.next;
    }
};

File: src/p0002/rust/solution.rs
impl Solution {
    pub fn add_two_numbers(p: Option<Box<ListNode>>, q: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut a, mut b, mut carry) = (p, q, 0);
        let mut sentinel = Box::new(ListNode::new(0));
        let mut tail = &mut sentinel;
        while carry != 0 || a.is_some() || b.is_some() {
            if let Some(t) = a {
                carry += t.val;
                a = t.next;
            }
            if let Some(t) = b {
                carry += t.val;
                b = t.next;
            }
            tail.next = Some(Box::new(ListNode::new(carry % 10)));
            tail = tail.next.as_mut().unwrap();
            carry /= 10;
        }
        sentinel.next
    }
}
